index.html

```diff
@@ -1,390 +1,1451 @@
+              <li>Invoke the operating system to discard the contents of all
+              software and hardware receive buffers for the port.
+              <li>Invoke the steps to [=handle closing the readable stream=].
+              <li>
+                [=Queue a task=] to resolve |promise| on the [=serial port task
+                source=].
+            </ol>
+          <li>Return |promise|.
+        </ol>
+
+      <li>
+        [=ReadableStream/Set up=] |stream| with
+        <a href="https://streams.spec.whatwg.org/#readablestream-set-up-pullalgorithm">pullAlgorithm</a> set
+        to |pullAlgorithm| and
+        <a href="https://streams.spec.whatwg.org/#readablestream-set-up-cancelalgorithm">cancelAlgorithm</a> set
+        to |cancelAlgorithm|.
+      <li>Set [=this=].{{[[readable]]}} to |stream|.
+      <li>Return |stream|.
+    </ol>
+
+    To <dfn>handle closing the readable stream</dfn> perform the following
+    steps:
+
+    <ol>
+      <li>Set [=this=].{{[[readable]]}} to `null`.
+      <li>
+        If [=this=].{{[[writable]]}} is `null` and
+        [=this=].{{[[pendingClosePromise]]}} is not `null`, [=queue a task=] to
+        [=resolve=] [=this=].{{[[pendingClosePromise]]}} with `undefined` on the
+        [=serial port task source=].
+    </ol>
+
+  </section>
+
+  <section>
+    <h3><dfn>writable</dfn> attribute</h3>
+
+    <aside class="example">
+      To write individual chunks of data to the port a
+      {{WritableStreamDefaultWriter}} can be created and released as necessary.
+      This example uses a `TextEncoder` to encode a {{DOMString}} as the
+      necessary {{Uint8Array}} for transmission.
+
+      <pre class="js">
+        const encoder = new TextEncoder();
+        const writer = port.writable.getWriter();
+        await writer.write(encoder.encode("PING"));
+        writer.releaseLock();
+      </pre>
+
+      When writing larger chunks it can be important to allow the port to apply
+      back pressure so that the serial transmitter does not get too far behind
+      sending data generated by the application. The
+      {{WritableStreamDefaultWriter/write()}} method returns a {{Promise}} which
+      resolves when data has been written. While having some data available in
+      the transmit buffer is important to maintain good throughput awaiting this
+      {{Promise}} before generating too many chunks of data is a good practice
+      to avoid excessive buffering.
+    </aside>
+
+    The {{SerialPort/writable}} getter steps are:
+
+    <ol>
+      <li>
+        If [=this=].{{[[writable]]}} is not `null`, return
+        [=this=].{{[[writable]]}}.
+      <li>If [=this=].{{[[state]]}} is not `"opened"`, return `null`.
+      <li>If [=this=].{{[[writeFatal]]}} is `true`, return `null`.
+      <li>Let |stream:WritableStream| be a [=new=] {{WritableStream}}.
+      <li>Let |writeAlgorithm| be the following steps:
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          <li>
```

Comment:
Probably before going in parallel, you need to handle what happens if chunk is not a Uint8Array. I.e. you should convert it to a [byte sequence](https://infra.spec.whatwg.org/#byte-sequence) in the synchronous section, returning a rejected promise if the conversion fails.

Reply:
In Blink this gets converted into an `ArrayBufferOrArrayBufferView` though it might be more correct to throw a `TypeError` if it is anything other than a `Uint8Array`. I don't want to support too many magical conversions. Is there an established algorithm for converting a value to a byte sequence?

Reply:
You probably want https://heycam.github.io/webidl/#dfn-get-buffer-source-copy, but that assumes you first have a spec-`BufferSource` (= Blink's `ArrayBufferOrArrayBufferView`). So I think first you should [convert](https://heycam.github.io/webidl/#dfn-convert-ecmascript-to-idl-value) to either a `BufferSource` or a `Uint8Array`.

I think `BufferSource` (instead of being strict on `Uint8Array`) is the right move for inputs. That's what `TextDecoderStream` and `CompressionStream` accept, and in general for web platform APIs we try to use `BufferSource` on the input side.